checker/src/environment.rs

```diff
@@ -162,17 +166,20 @@ impl Environment {
                 // abstract interpretation because the loop exit condition may evaluate to false
                 // for the first few iterations of the loop and thus the backwards branch only becomes
                 // conditional once widening kicks in.
-                x.join(y, p)
+                x.clone().join(y.clone(), p.clone())
```

Comment:
Same here, the x has to be cloned? Your diff to clone less has three clones on one line :)

Reply:
Before this PR, join would do the cloning. Now the caller gets to decide. Sometimes, the caller knows that the value will not be reused. That causes fewer clones to happen at runtime.